File: intern/633.sum-of-square-numbers.js
/*
 * @lc app=leetcode id=633 lang=javascript
 *
 * [633] Sum of Square Numbers
 */

// @lc code=start
/**
 * @param {number} c
 * @return {boolean}
 */
var judgeSquareSum = function (c) {
    let a = 0;
    let b = Math.floor(Math.sqrt(c));

    let aSquare = a * a;
    let bSquare = b * b;

    while (a <= b) {
        const squareSum = aSquare + bSquare;
        if (squareSum < c) {
            a++;
            aSquare = a * a;
        } else if (squareSum > c) {
            b--;
            bSquare = b * b;
        } else {
            return true;
        }
    }

    return false;
};
// @lc code=end

console.log(judgeSquareSum(0));

File: prev-2024/633.sum-of-square-numbers.cpp
/*
 * @lc app=leetcode id=633 lang=cpp
 *
 * [633] Sum of Square Numbers
 */

#include <cinttypes>
#include <cmath>

// @lc code=start
class Solution {
 public:
  bool judgeSquareSum(int32_t c) {
    std::int32_t left = 0;
    std::int32_t right = static_cast<std::int32_t>(std::sqrt(c));

    while (left <= right) {
      std::int64_t mid = static_cast<std::int64_t>(left * left) +
                         static_cast<std::int64_t>(right * right);
      if (mid < c) {
        left++;
      } else if (mid > c) {
        right--;
      } else {
        return true;
      }
    }

    return false;
  }
};
// @lc code=end
